File: packages/a2a-py/src/grantex_a2a/_agent_card.py
"""Agent Card builder with Grantex authentication extensions."""

from __future__ import annotations

from ._types import A2AAgentCard, GrantexAgentCardOptions


def build_grantex_agent_card(options: GrantexAgentCardOptions) -> A2AAgentCard:
    """Build an A2A agent card with Grantex authentication configured.

    Args:
        options: Agent card configuration.

    Returns:
        An A2AAgentCard with Grantex auth scheme.

    Example::

        card = build_grantex_agent_card(GrantexAgentCardOptions(
            name="My Agent",
            description="An agent that does things",
            url="https://my-agent.example.com/a2a",
            jwks_uri="https://grantex.dev/.well-known/jwks.json",
            issuer="https://grantex.dev",
            required_scopes=["read", "write"],
            delegation_allowed=True,
        ))
    """
    grantex_config = {
        "jwksUri": options.jwks_uri,
        "issuer": options.issuer,
    }
    if options.required_scopes is not None:
        grantex_config["requiredScopes"] = options.required_scopes  # type: ignore[assignment]
    if options.delegation_allowed is not None:
        grantex_config["delegationAllowed"] = options.delegation_allowed  # type: ignore[assignment]

    return A2AAgentCard(
        name=options.name,
        description=options.description,
        url=options.url,
        version=options.version,
        provider=options.provider,
        capabilities=options.capabilities,
        authentication={
            "schemes": [
                {
                    "scheme": "bearer",
                    "grantexConfig": grantex_config,
                }
            ]
        },
        skills=options.skills,
    )

File: packages/a2a/src/agent-card.ts
/**
 * Agent Card builder with Grantex authentication extensions.
 *
 * Generates A2A-compliant agent cards with Grantex auth configuration
 * for grant token validation.
 */

import type { A2AAgentCard, GrantexAgentCardOptions } from './types.js';

/**
 * Build an A2A agent card with Grantex authentication configured.
 *
 * @example
 * ```ts
 * const card = buildGrantexAgentCard({
 *   name: 'My Agent',
 *   description: 'An agent that does things',
 *   url: 'https://my-agent.example.com/a2a',
 *   jwksUri: 'https://grantex.dev/.well-known/jwks.json',
 *   issuer: 'https://grantex.dev',
 *   requiredScopes: ['read', 'write'],
 *   delegationAllowed: true,
 * });
 * ```
 */
export function buildGrantexAgentCard(options: GrantexAgentCardOptions): A2AAgentCard {
  return {
    name: options.name,
    description: options.description,
    url: options.url,
    ...(options.version !== undefined ? { version: options.version } : {}),
    ...(options.provider !== undefined ? { provider: options.provider } : {}),
    ...(options.capabilities !== undefined ? { capabilities: options.capabilities } : {}),
    authentication: {
      schemes: [
        {
          scheme: 'bearer',
          grantexConfig: {
            jwksUri: options.jwksUri,
            issuer: options.issuer,
            ...(options.requiredScopes !== undefined ? { requiredScopes: options.requiredScopes } : {}),
            ...(options.delegationAllowed !== undefined ? { delegationAllowed: options.delegationAllowed } : {}),
          },
        },
      ],
    },
    defaultInputModes: ['text/plain'],
    defaultOutputModes: ['text/plain'],
    ...(options.skills !== undefined ? { skills: options.skills } : {}),
  };
}
